### PR TITLE
Surround detection

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

# Description

implement detection logic of min max span detection method from here https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround

### CODE REVIEW

File: slasher/db/indexed_attestations.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	"reflect"
 	"sort"
 
+	"github.com/prysmaticlabs/prysm/shared/params"
+
```
Comment (rauljordan) [inline slasher/db/indexed_attestations.go]:
```suggestion
```

File: slasher/rpc/detect_update_min_max_span_bench_test.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	defer db.TeardownSlasherDB(b, dbs)
+	ctx := context.Background()
+	slasherServer := &Server{
+		SlasherDB: dbs,
+	}
+	for _, diff := range diffs {
+		b.Run(fmt.Sprintf("MinSpan with diff: %d", diff), func(ib *testing.B) {
+			for i := uint64(ib.N) - 10; i < uint64(ib.N); i++ {
+				_, _ = slasherServer.DetectAndUpdateMinSpan(ctx, i, i+diff, 1)
+
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span_bench_test.go]:
```suggestion
```

File: slasher/rpc/detect_update_min_max_span_bench_test.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+func BenchmarkMinSpan(b *testing.B) {
+	diffs := []uint64{2, 10, 100, 1000, 10000, 53999}
+	dbs := db.SetupSlasherDB(b)
+	defer db.TeardownSlasherDB(b, dbs)
+	ctx := context.Background()
+	slasherServer := &Server{
+		SlasherDB: dbs,
+	}
+	for _, diff := range diffs {
+		b.Run(fmt.Sprintf("MinSpan with diff: %d", diff), func(ib *testing.B) {
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span_bench_test.go]:
```suggestion
		b.Run(fmt.Sprintf("MinSpan_diff_%d", diff), func(ib *testing.B) {
```

File: slasher/rpc/detect_update_min_max_span_bench_test.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+			for i := uint64(ib.N) - 10; i < uint64(ib.N); i++ {
+				_, _ = slasherServer.DetectAndUpdateMinSpan(ctx, i, i+diff, 1)
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span_bench_test.go]:
```suggestion
				slasherServer.DetectAndUpdateMinSpan(ctx, i, i+diff, 1)
```

File: slasher/rpc/detect_update_min_max_span_test.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	slasherServer := &Server{
+		SlasherDB: dbs,
+	}
+	for _, tt := range spanTestsMax {
+		st, err := slasherServer.DetectAndUpdateMaxSpan(ctx, tt.sourceEpoch, tt.targetEpoch, tt.validatorIdx)
+		if err != nil {
+			t.Fatalf("Failed to update span: %v", err)
+		}
+		if st != tt.slashingTargetEpoch {
+			t.Fatalf("Expected slashing target : %v got: %v", tt.slashingTargetEpoch, st)
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span_test.go]:
```suggestion
			t.Fatalf("Expected slashing target: %v got: %v", tt.slashingTargetEpoch, st)
```

File: slasher/rpc/detect_update_min_max_span_test.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+	slasherServer := &Server{
+		SlasherDB: dbs,
+	}
+	for _, tt := range spanTestsMin {
+		st, err := slasherServer.DetectAndUpdateMinSpan(ctx, tt.sourceEpoch, tt.targetEpoch, tt.validatorIdx)
+		if err != nil {
+			t.Fatalf("failed to update span: %v", err)
+		}
+		if st != tt.slashingTargetEpoch {
+			t.Fatalf("Expected slashing target : %v got: %v", tt.slashingTargetEpoch, st)
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span_test.go]:
```suggestion
			t.Fatalf("Expected slashing target: %v got: %v", tt.slashingTargetEpoch, st)
```

File: slasher/rpc/detect_update_min_max_span_test.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+func TestServer_FailToUpdate(t *testing.T) {
+	dbs := db.SetupSlasherDB(t)
+	defer db.TeardownSlasherDB(t, dbs)
+	ctx := context.Background()
+	slasherServer := &Server{
+		SlasherDB: dbs,
+	}
+	spanTestsFail := spanMapTestStruct{
+
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span_test.go]:
```suggestion
```

File: slasher/db/indexed_attestations.go [hunk 8]
```diff
@@ -117,7 +119,7 @@ func (db *Store) SaveIndexedAttestation(idxAttestation *ethpb.IndexedAttestation
 
 	// prune history to max size every 10th epoch
```
Comment (rauljordan) [inline slasher/db/indexed_attestations.go]:
```suggestion
	// prune history to max size every 10th epoch.
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// DetectAndUpdateMaxSpan is used to detect and update the max span of an incoming attestation.
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+func (ss *Server) DetectAndUpdateMaxSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (surroundedTargetEpoch uint64, err error) {
+	span := target - source + 1
+	if span > params.BeaconConfig().WeakSubjectivityPeriod {
+		return 0, fmt.Errorf("attestation detection supports only weak subjectivity period: %v target - source: %v > weakSubjectivityPeriod", params.BeaconConfig().WeakSubjectivityPeriod, span)
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: Line too long, make multiline as:
```suggestion
		return 0, fmt.Errorf("
		    %v target - source: %v > weakSubjectivityPeriod", 
		    params.BeaconConfig().WeakSubjectivityPeriod, 
		    span,
		)
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+	}
+	spanMap, err := ss.SlasherDB.ValidatorSpansMap(validatorIdx)
+	if err != nil {
+		return 0, errors.Wrapf(err, "could not retrieve span map for validatorIdx: %v", validatorIdx)
+	}
+	if spanMap.EpochSpanMap == nil {
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: spanMap itself could be nil, no? That should also be checked. Maybe we can also do:

```go
if _, ok := spanMap.EpochSpanMap[source]; ok {
}
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+			maxSpan := uint64(spanMap.EpochSpanMap[source].MaxSpan)
+			if maxSpan > span {
+				return maxSpan + source, nil
+			}
+		}
+	}
+
+	for i := uint64(1); i < target-source; i++ {
+		val := uint32(span - i - 1)
+		if spanMap.EpochSpanMap[source+i] == nil {
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: instead do `if _, ok := spanMap.EpochSpanMap[source+i]; !ok {`

File: slasher/rpc/detect_update_min_max_span.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+// DetectAndUpdateMinSpan is used to detect surround and update the min span
+// of an incoming attestation.
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+func (ss *Server) DetectAndUpdateMinSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (surroundTargetEpoch uint64, err error) {
+	span := target - source + 1
+	if span > params.BeaconConfig().WeakSubjectivityPeriod {
+		return 0, fmt.Errorf("attestation slashing detection supports only weak subjectivity period: %v target - source: %v > weakSubjectivityPeriod", params.BeaconConfig().WeakSubjectivityPeriod, span)
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: +1 as above. Also, this logic is duplicated in the max span function, can this be abstracted away into something common?

File: slasher/rpc/detect_update_min_max_span.go [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+	}
+	spanMap, err := ss.SlasherDB.ValidatorSpansMap(validatorIdx)
+	if err != nil {
+		return 0, errors.Wrapf(err, "could not retrieve span map for validatorIdx: %v", validatorIdx)
+	}
+	if spanMap.EpochSpanMap == nil {
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: +1 on duplication here

File: slasher/rpc/detect_update_min_max_span.go (hunk 11 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: thanks, changed

File: slasher/rpc/detect_update_min_max_span.go (hunk 13 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: thanks. changed

File: slasher/rpc/detect_update_min_max_span.go (hunk 12 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: thanks! changed

File: slasher/rpc/detect_update_min_max_span.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+type detect func(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64
```
Comment (prestonvanloon) [inline slasher/rpc/detect_update_min_max_span.go]: Can you elaborate on this type? Maybe a better name, it's not clear what this is used for.

File: slasher/rpc/detect_update_min_max_span.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+	if _, ok := spanMap.EpochSpanMap[source]; ok {
+		return detectionFunc(span, spanMap.EpochSpanMap[source], source), span, spanMap, nil
+	}
+	return 0, span, spanMap, nil
+}
+
+// DetectAndUpdateMaxSpan is used to detect and update the max span of an incoming attestation.
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+func (ss *Server) DetectAndUpdateMaxSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (surroundedTargetEpoch uint64, err error) {
```
Comment (prestonvanloon) [inline slasher/rpc/detect_update_min_max_span.go]: Please don't use named return variables. 

```suggestion
func (ss *Server) DetectAndUpdateMaxSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (uint64, error) {
```

Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: removing. thanks

File: slasher/rpc/detect_update_min_max_span.go [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+	"fmt"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+type detect func(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64
+
+func detectMax(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64 {
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: Would recommend putting unexported funcs at the bottom of the file if possible, especially if they don't have much context or are helper functions

File: slasher/rpc/detect_update_min_max_span.go [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+func detectMin(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64 {
+	minSpan := uint64(spans.MinSpan)
+	if minSpan < span {
+		return minSpan + source
+	}
+	return 0
+}
+
+func (ss *Server) detectSpan(source, target, validatorIdx uint64, detectionFunc detect) (uint64, uint64, *ethpb.EpochSpanMap, error) {
+	span := target - source + 1
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: To an inexperienced reader, it isn't clear exactly what a `span` is. I understand it as a span of epochs between target and source, but we should be explaining some basic information about what a target and source are I think at least in the godocs of the `DetectAndUpdateMaxSpan` or min span exported functions. Maybe we could rename this variable to `epochSpan`? or some other measure to denote it is a distance metric?

Comment (rauljordan) [reply slasher/rpc/detect_update_min_max_span.go]: Just mostly reviewing this from the standpoint of readability and maintainability.

File: slasher/rpc/detect_update_min_max_span.go [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+func detectMin(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64 {
+	minSpan := uint64(spans.MinSpan)
+	if minSpan < span {
+		return minSpan + source
+	}
+	return 0
+}
+
+func (ss *Server) detectSpan(source, target, validatorIdx uint64, detectionFunc detect) (uint64, uint64, *ethpb.EpochSpanMap, error) {
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: This func could really benefit from some basic comments even if it is unexported, maybe explaining that how it takes in a callback that does detection.

File: slasher/rpc/detect_update_min_max_span.go [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+			break
+		}
+	}
+	if err := ss.SlasherDB.SaveValidatorSpansMap(validatorIdx, spanMap); err != nil {
+		return 0, err
+	}
+	return 0, nil
+}
+
+// DetectAndUpdateMinSpan is used to detect surround and update the min span
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: +1 as above, maybe we could explain some of these terminologies in simpler terms for someone who isn't an expert in slashing

File: slasher/rpc/detect_update_min_max_span.go (hunk 17 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: thanks, epochSpan sounds better

Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: will add a comment

File: slasher/rpc/detect_update_min_max_span.go (hunk 18 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: adding

File: slasher/rpc/detect_update_min_max_span.go (hunk 14 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: i added a comment and changed naming. please tell me if it makes sense

File: slasher/rpc/detect_update_min_max_span.go (hunk 16 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: done

File: slasher/rpc/detect_update_min_max_span.go (hunk 17 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: can you please take a look? i added comment and name changed

File: slasher/rpc/detect_update_min_max_span.go [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+type detect func(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: Please add more detail on this type and why it is needed, typically this would be a better fit for something like an interface like:
```go
type Detector interface {
  Detect()
}
```
and create a struct that matches that interface. Then, the interface can be an input argument anywhere it is needed, allowing for a nice abstraction over detection itself. This interface pattern is more idiomatic with Go.

File: slasher/rpc/detect_update_min_max_span.go [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+
+// DetectAndUpdateMaxEpochSpan is used to detect and update the max span of an incoming attestation.
+// max span is the span between the current attestation source epoch and the furthest attestation
+// target that its source epoch is lower then it.
+// distance.
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
//
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+	if err := ss.SlasherDB.SaveValidatorSpansMap(validatorIdx, spanMap); err != nil {
+		return 0, err
+	}
+	return 0, nil
+}
+
+// DetectAndUpdateMinEpochSpan is used to detect surround and update the min epoch span
+// of an incoming attestation.
+// min span is the span between the current attestation and the closest attestation target
+// distance.
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
// distance.
//
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+		return minSpan + attestationSourceEpoch
+	}
+	return 0
+}
+
+// detectSlashingByEpochSpan is used to detect if a slashable event is present
+// in the db by checking either the closest attestation target or the furthest
+// attestation target. this method receives the detection func in order to be
+// compatible for both cases.
+func (ss *Server) detectSlashingByEpochSpan(source, target, validatorIdx uint64, detectionFunc detect) (uint64, uint64, *ethpb.EpochSpanMap, error) {
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]: See comment from earlier, I'd recommend passing in a detector interface into this func and having structs that implement such interface

File: proto/eth/v1alpha1/slasher.proto [hunk 24]
```diff
@@ -51,13 +51,15 @@ message ProposerSlashingResponse {
...
-message MinMaxSpan {
-    uint32 min_span = 1;
-    uint32 max_span = 2;
+message MinMaxEpochSpan {
+    uint32 min_epoch_span = 1;
+    uint32 max_epoch_span = 2;
 }
-// every validator will have his own spans map containing min and max value for each epoch
-// in order to detect slashable attestation as quick as possible.
+// Every validator will have his own spans map containing min distance from each epoch
```
Comment (0xKiwi) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
// Every validator will have their own spans map containing min distance from each epoch
```

File: proto/eth/v1alpha1/slasher.proto [hunk 25]
```diff
@@ -51,13 +51,15 @@ message ProposerSlashingResponse {
 // each attestation source to those spans
 // see https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
 // for further details.
-message MinMaxSpan {
-    uint32 min_span = 1;
-    uint32 max_span = 2;
+message MinMaxEpochSpan {
+    uint32 min_epoch_span = 1;
```
Comment (nisdas) [inline proto/eth/v1alpha1/slasher.proto]: whats the reason for `uint32` instead of  `uint64` ?

File: slasher/rpc/detect_update_min_max_span.go [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+	Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64
+}
+
+// maxDetector is a detector used to detect surround attestations.
+type maxDetector struct{}
+
+// minDetector is a detector used to detect surrounded attestations.
+type minDetector struct{}
+
+func (d maxDetector) Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64 {
```
Comment (nisdas) [inline slasher/rpc/detect_update_min_max_span.go]: can you put the method after the type declaration ? Having two methods with the same name is confusing one after the other

File: slasher/rpc/detect_update_min_max_span_bench_test.go [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+	defer db.TeardownSlasherDB(b, dbs)
+
+	ctx := context.Background()
+	slasherServer := &Server{
+		SlasherDB: dbs,
+	}
+	for _, diff := range diffs {
+		b.Run(fmt.Sprintf("MaxSpan_diff_%d", diff), func(ib *testing.B) {
+			for i := uint64(0); i < uint64(ib.N); i++ {
+				_, err := slasherServer.DetectAndUpdateMaxEpochSpan(ctx, diff, diff+i, i%10)
```
Comment (nisdas) [inline slasher/rpc/detect_update_min_max_span_bench_test.go]: whats the `i%10` for ?

File: slasher/rpc/detect_update_min_max_span_bench_test.go [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+	dbs := db.SetupSlasherDB(b)
+	defer db.TeardownSlasherDB(b, dbs)
+
+	slasherServer := &Server{
+		SlasherDB: dbs,
+	}
+	for _, diff := range diffs {
+		b.Run(fmt.Sprintf("Detect_MaxSpan_diff_%d", diff), func(ib *testing.B) {
+			for i := uint64(0); i < uint64(ib.N); i++ {
+				_, _, _, err := slasherServer.detectSlashingByEpochSpan(i, i+diff, i%10, detectEpochMaxSpan)
```
Comment (nisdas) [inline slasher/rpc/detect_update_min_max_span_bench_test.go]: same here

File: proto/eth/v1alpha1/slasher.proto (hunk 25 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: less space. i would be good with 16

File: slasher/rpc/detect_update_min_max_span_bench_test.go (hunk 27 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span_bench_test.go]: to use few validators indices and have multiple updates on each of the maps

File: slasher/rpc/detect_update_min_max_span_bench_test.go (hunk 28 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span_bench_test.go]: same answer

File: slasher/rpc/detect_update_min_max_span.go (hunk 26 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: thanks, changed

File: slasher/rpc/detect_update_min_max_span.go [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// Detector is an interface used to implement the slashing surround an surrounded detection
+// methods.
+type Detector interface {
+	Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
	Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, sourceEpoch uint64) uint64
```

File: proto/eth/v1alpha1/slasher.proto [hunk 30]
```diff
@@ -51,13 +51,15 @@ message ProposerSlashingResponse {
...
+    uint32 max_epoch_span = 2;
 }
-// every validator will have his own spans map containing min and max value for each epoch
-// in order to detect slashable attestation as quick as possible.
+// Every validator will have their own spans map containing min distance from each epoch
+// to the closest target epoch of another attestation (surrounded) and max distance to
+// a target attestation (surrounding), in order to detect slashable attestation as quick
```
Comment (terencechain) [inline proto/eth/v1alpha1/slasher.proto]:
```suggestion
// a target attestation (surrounding), in order to detect slashable attestation as quickly
```

File: slasher/rpc/detect_update_min_max_span.go (hunk 29 above)
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
	Detect(attestationEpochSpan uint64, recorderEpochSpan *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+// maxDetector is a detector used to detect surround attestations.
+type maxDetector struct{}
+
+// Detect max detector is a detector used to detect surround attestations.
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
// Detect is used to detect surround attestations.
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 32]
```diff
@@ -0,0 +1,135 @@
...
+func (d maxDetector) Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64 {
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]: this line is too long

File: slasher/rpc/detect_update_min_max_span.go [hunk 33]
```diff
@@ -0,0 +1,135 @@
...
+	if minSpan < attestationEpochSpan {
+		return minSpan + attestationSourceEpoch
+	}
+	return 0
+}
+
+// DetectAndUpdateMaxEpochSpan is used to detect and update the max span of an incoming attestation.
+// max span is the span between the current attestation source epoch and the furthest attestation
+// target that its source epoch is lower then it.
+//
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 34]
```diff
@@ -0,0 +1,135 @@
...
+func (ss *Server) detectSlashingByEpochSpan(source, target, validatorIdx uint64, detector interface{}) (uint64, uint64, *ethpb.EpochSpanMap, error) {
+	d, ok := detector.(Detector)
+	if !ok {
+		return 0, 0, nil, fmt.Errorf("detector interface should be of type Detector. got: %v",
+			reflect.TypeOf(detector),
+		)
+	}
+	span := target - source + 1
+	if span > params.BeaconConfig().WeakSubjectivityPeriod {
+		return 0, span, nil, fmt.Errorf("%d target - source: %d > weakSubjectivityPeriod",
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
		return 0, span, nil, fmt.Errorf("target: "%d - source: %d > weakSubjectivityPeriod",
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 35]
```diff
@@ -0,0 +1,135 @@
...
+			params.BeaconConfig().WeakSubjectivityPeriod,
+			span,
+		)
+	}
+	spanMap, err := ss.SlasherDB.ValidatorSpansMap(validatorIdx)
+	if err != nil {
+		return 0, span, nil, errors.Wrapf(err, "could not retrieve span map for validatorIdx: %d", validatorIdx)
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
		return 0, span, nil, errors.Wrapf(err, "could not retrieve span map for validator index: %d", validatorIdx)
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 36]
```diff
@@ -0,0 +1,135 @@
...
+		return targetEpoch, nil
+	}
+	for i := uint64(1); i < target-source; i++ {
+		val := uint32(span - i - 1)
+		if _, ok := spanMap.EpochSpanMap[source+i]; !ok {
+			spanMap.EpochSpanMap[source+i] = &ethpb.MinMaxEpochSpan{}
+		}
+		if spanMap.EpochSpanMap[source+i].MaxEpochSpan < val {
+			spanMap.EpochSpanMap[source+i].MaxEpochSpan = val
+		} else {
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]: I dont think you need this else

File: slasher/rpc/detect_update_min_max_span.go [hunk 37]
```diff
@@ -0,0 +1,135 @@
...
+		return 0, nil
+	}
+	for i := source - 1; i > 0; i-- {
+		val := uint32(target - (i))
+		if _, ok := spanMap.EpochSpanMap[i]; !ok {
+			spanMap.EpochSpanMap[i] = &ethpb.MinMaxEpochSpan{}
+		}
+		if spanMap.EpochSpanMap[i].MinEpochSpan == 0 || spanMap.EpochSpanMap[i].MinEpochSpan > val {
+			spanMap.EpochSpanMap[i].MinEpochSpan = val
+		} else {
```
Comment (terencechain) [inline slasher/rpc/detect_update_min_max_span.go]: I dont think you need this else

File: slasher/rpc/detect_update_min_max_span.go [hunk 38]
```diff
@@ -0,0 +1,134 @@
...
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// Detector is an interface used to implement the slashing surround an surrounded detection
+// methods.
+type Detector interface {
+	Detect(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, sourceEpoch uint64) uint64
```
Comment (rauljordan) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
	Detect(attestationEpochSpan uint64, recorderEpochSpan *ethpb.MinMaxEpochSpan, sourceEpoch uint64) uint64
```

File: proto/eth/v1alpha1/slasher.proto (hunk 25 above)
Comment (0xKiwi) [reply proto/eth/v1alpha1/slasher.proto]: Why not use 16?

File: slasher/rpc/detect_update_min_max_span.go [hunk 39]
```diff
@@ -0,0 +1,134 @@
...
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// Detector is an interface used to implement the slashing surround an surrounded detection
```
Comment (0xKiwi) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
// Detector is an interface used to implement the slashable surrounded/surrounding vote detection
```

File: slasher/rpc/detect_update_min_max_span.go [hunk 40]
```diff
@@ -0,0 +1,134 @@
...
+		return minSpan + attestationSourceEpoch
+	}
+	return 0
+}
+
+// DetectAndUpdateMaxEpochSpan is used to detect and update the max span of an incoming attestation.
+// max span is the span between the current attestation source epoch and the furthest attestation
+// target that its source epoch is lower then it.
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
```
Comment (0xKiwi) [inline slasher/rpc/detect_update_min_max_span.go]: Please capitalize the first letters of each sentence here

File: slasher/rpc/detect_update_min_max_span.go [hunk 41]
```diff
@@ -0,0 +1,134 @@
...
+	}
+	return 0, nil
+}
+
+// DetectAndUpdateMinEpochSpan is used to detect surround and update the min epoch span
+// of an incoming attestation.
+// min span is the span between the current attestation and the closest attestation target
+// distance.
+//
+// logic is following the detection method designed by https://github.com/protolambda
```
Comment (0xKiwi) [inline slasher/rpc/detect_update_min_max_span.go]: Capitalize here too please

File: slasher/rpc/detect_update_min_max_span.go [hunk 42]
```diff
@@ -0,0 +1,134 @@
...
+	}
+	if err := ss.SlasherDB.SaveValidatorSpansMap(validatorIdx, spanMap); err != nil {
+		return 0, errors.Wrap(err, "could not save validator spans")
+	}
+	return 0, nil
+}
+
+// detectSlashingByEpochSpan is used to detect if a slashable event is present
+// in the db by checking either the closest attestation target or the furthest
+// attestation target. this method receives the detection func in order to be
```
Comment (0xKiwi) [inline slasher/rpc/detect_update_min_max_span.go]:
```suggestion
// attestation target. This method receives the detection func in order to be
```

File: slasher/rpc/detect_update_min_max_span.go (hunk 32 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: fixed. thanks

File: slasher/rpc/detect_update_min_max_span.go (hunk 36 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: i want to stop updating values if i find equal or higher values

File: slasher/rpc/detect_update_min_max_span.go (hunk 37 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: why not

File: slasher/rpc/detect_update_min_max_span.go (hunk 40 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: thanks

File: proto/eth/v1alpha1/slasher.proto (hunk 25 above)
Comment (shayzluf) [reply proto/eth/v1alpha1/slasher.proto]: sadly protobuf dont have uint16

File: slasher/rpc/detect_update_min_max_span.go (hunk 36 above)
Comment (terencechain) [reply slasher/rpc/detect_update_min_max_span.go]: I see it now!

File: slasher/rpc/detect_update_min_max_span.go (hunk 37 above)
Comment (shayzluf) [reply slasher/rpc/detect_update_min_max_span.go]: same reason as above

File: slasher/rpc/detect_update_min_max_span.go [hunk 43]
```diff
@@ -0,0 +1,142 @@
...
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// Detector is an interface used to implement the slashable surrounding/surrounded
+// vote detection methods.
+type Detector interface {
+	Detect(attestationEpochSpan uint64, recorderEpochSpan *ethpb.MinMaxEpochSpan, sourceEpoch uint64) uint64
```
Comment (prestonvanloon) [inline slasher/rpc/detect_update_min_max_span.go]: Instead of defining an interface, can you define a function type? this is a bit cumbersome to use. 

```go
type DetectFn = func(attestationEpochSpan uint64, recorderEpochSpan *ethpb.MinMaxEpochSpan, sourceEpoch uint64) uint64
```

Then you can change the API of detectSlashingByEpochSpan
```go
func detectSlashingByEpochSpan(source, target, validatorIdx uint64, detect DetectFn) {
  if detect(...) > 0 {
     // do something
   }
}
```
